addons/stock/models/stock_replenish_mixin.py

```diff
@@ -26,11 +26,13 @@ def _compute_allowed_route_ids(self):
     def _get_allowed_route_domain(self):
         stock_location_inter_company_id = self.env.ref('stock.stock_location_inter_company').id
 
-        base_domain = Domain('product_selectable', '=', True)
-        if self.warehouse_id:
-            wh_route_ids = self.warehouse_id.route_ids.filtered(lambda r: r._is_valid_resupply_route_for_product(self.product_id)).ids
+        base_domain = []
+        warehouse = self.warehouse_id or self.env['stock.warehouse'].browse(self.env.context.get('warehouse_id', False))
+        if warehouse:
+            base_domain = Domain.AND([base_domain, Domain('rule_ids.location_dest_id.warehouse_id', 'in', warehouse.ids), Domain('rule_ids.action', 'not in', ['buy', 'manufacture'])])
+            wh_route_ids = warehouse.route_ids.filtered(lambda r: r._is_valid_resupply_route_for_product(self.product_id)).ids
             if wh_route_ids:
-                base_domain |= Domain('id', 'in', wh_route_ids)
+                base_domain = Domain.OR([Domain('id', 'in', wh_route_ids), base_domain])
```

Comment:
```suggestion
                domain |= Domain(('id', 'in', wh_route_ids))
```

This is simpler, and I would use the domain instead of the base_domain.
It keeps the code cleaner and easier to maintain in the long run.